Authentication polish

shrralis opened the pull request:
Fixed bug with logout when it just didn't work and added new security logic on authentication and password recovering.

Need a suggestion with Jackson config. (remove unnecessary `@Bean` in `AppConfig.java` or what)

shrralis committed:
# Shrralis commit changelog #37:

##    [GLOBAL]:
     - Added some new security logic.

##    [Front-end]:
     - Fixed bug with logout.
     - Added latest error constants.
     - New `md-chip` styles.

##    [Back-end]:
     - Fixed bug with logout.
     - Added new error constants.
     - New checking on failed authentication attempts.
     - New checking on "ddos" with creating recovery tokens.

##    [Database]:
	 - New column in table `users`, that is `failed_auth_count`.

shrralis committed:
# Shrralis commit changelog #37:

##    [GLOBAL]:
     - Added some new security logic.

##    [Front-end]:
     - Fixed bug with logout.
     - Added latest error constants.
     - New `md-chip` styles.

##    [Back-end]:
     - Fixed bug with logout.
     - Added new error constants.
     - New checking on failed authentication attempts.
     - New checking on "ddos" with creating recovery tokens.

##    [Database]:
	 - New column in table `users`, that is `failed_auth_count`.

shrralis committed:
# Shrralis commit changelog #37:

##    [GLOBAL]:
     - Added new README.

##    [Front-end]:
     `null`

##    [Back-end]:
     `null`

##    [Database]:
	 `null`

shrralis committed:
# Shrralis commit changelog #38:

##    [GLOBAL]:
     `null`

##    [Front-end]:
     - Fixed URL for getting current session user info.

##    [Back-end]:
     - Added new RestController mapping for getting current session user info.

##    [Database]:
	 `null`

shrralis commented on src/main/java/com/shrralis/ssdemo1/repository/RecoveryTokensRepository.java:
```diff
@@ -17,4 +18,9 @@
+
+	@Query(value = "SELECT COUNT(*) FROM " + RecoveryToken.TABLE_NAME +
+			" WHERE " + RecoveryToken.USER_COLUMN_NAME + " = ?1 AND " +
+			RecoveryToken.EXPIRES_AT_COLUMN_NAME + " >= NOW() ", nativeQuery = true)
```
Please, don't be afraid of this

Roma7-7-7 replied:
There are no need to use constants. We are in repository for that entity so it's ok to just hardcode it

shrralis replied:
O.K., I'll replace it with just a string query in future

Roma7-7-7 commented on src/main/java/com/shrralis/ssdemo1/security/AuthProvider.java:
```diff
@@ -70,18 +82,21 @@ private UserDetails retrieveUserDetails(String login) {
             }
         } else {
-            throw new BadCredentialsException("Bad credentials");
+	        throw new BadCredentialsException(userDetails.getUsername());
```
Inverse this if.
if ( == null) { BadCredentialsException}
...

shrralis replied:
O.K., I missed that, thanks

Roma7-7-7 commented on src/main/java/com/shrralis/ssdemo1/security/handler/CitizenAuthenticationFailureHandler.java:
```diff
@@ -32,9 +36,22 @@ public void onAuthenticationFailure(HttpServletRequest request,
 		if (e.getClass().isAssignableFrom(UsernameNotFoundException.class)) {
 			logger.error("UsernameNotFoundException: {}", e);
 			MAPPER.writeValue(response.getWriter(), new JsonResponse(JsonError.Error.USER_NOT_EXIST.forField("login")));
+		} else if (e.getClass().isAssignableFrom(EmailNotFoundException.class)) {
```
Look at strategy and factory method patterns combination

shrralis replied:
I've changed that, look in coming commit, please.

shrralis committed:
# Shrralis commit changelog #38:

##    [GLOBAL]:
     - Applied Roma's suggestions.

##    [Front-end]:
     `null`

##    [Back-end]:
     `null`

##    [Database]:
	 `null`
